File: benchmark/python/benchmark.py
import timeit
import json
import msgpack
import sys
import os
import time
import struct

# Add local generated path for Proto/FB
sys.path.append(os.path.abspath("generated"))

from bench_complex import WorldState, Guild, Character, Item, Vec3

# Protobuf
from bench_complex_pb2 import (
    WorldState as ProtoWorldState,
    Guild as ProtoGuild,
    Character as ProtoCharacter,
    Item as ProtoItem,
    Vec3 as ProtoVec3,
)

# FlatBuffers
import flatbuffers
from bench_fb import WorldState as FBWorldState, Guild as FBGuild, Character as FBCharacter, Item as FBItem, Vec3 as FBVec3


def create_benchmark_data():
    world = WorldState()
    world.world_id = 1
    world.seed = "benchmark_seed"
    
    for g in range(1000):
        guild = Guild()
        guild.name = f"Guild_{g}"
        guild.description = "A very powerful guild"
        
        for c in range(20):
            char = Character()
            char.name = f"Char_{g}_{c}"
            char.level = c + 1
            char.hp = 100
            char.mp = 50
            char.is_alive = True
            
            pos = Vec3()
            pos.x = 10
            pos.y = 20
            pos.z = 30
            char.position = pos
            
            char.skills = [1, 2, 3, 4, 5]
            
            for i in range(10):
                item = Item()
                item.id = g * 1000 + c * 100 + i
                item.name = "Item" * 50 + f"_{g}_{c}_{i}"
                item.value = i * 10
                item.weight = 1
                item.rarity = "Rare" if i % 5 == 0 else "Common"
                char.inventory.append(item)
            
            guild.members.append(char)
        
        world.guilds.append(guild)
    
    world.loot_table = []
    return world


def to_dict(world):
    return {
        "world_id": world.world_id,
        "seed": world.seed,
        "guilds": [{
            "name": g.name,
            "description": g.description,
            "members": [{
                "name": c.name,
                "level": c.level,
                "hp": c.hp,
                "mp": c.mp,
                "is_alive": c.is_alive,
                "position": {"x": c.position.x, "y": c.position.y, "z": c.position.z},
                "skills": c.skills,
                "inventory": [{
                    "id": item.id,
                    "name": item.name,
                    "value": item.value,
                    "weight": item.weight,
                    "rarity": item.rarity,
                } for item in c.inventory]
            } for c in g.members]
        } for g in world.guilds],
        "loot_table": [],
    }


def create_proto_data(world):
    pw = ProtoWorldState()
    pw.world_id = world.world_id
    pw.seed = world.seed
    
    for g in world.guilds:
        pg = ProtoGuild()
        pg.name = g.name
        pg.description = g.description
        
        for c in g.members:
            pc = ProtoCharacter()
            pc.name = c.name
            pc.level = c.level
            pc.hp = c.hp
            pc.mp = c.mp
            pc.is_alive = c.is_alive
            
            pv = ProtoVec3()
            pv.x = c.position.x
            pv.y = c.position.y
            pv.z = c.position.z
            pc.position.CopyFrom(pv)
            
            pc.skills.extend(c.skills)
            
            for item in c.inventory:
                pi = ProtoItem()
                pi.id = item.id
                pi.name = item.name
                pi.value = item.value
                pi.weight = item.weight
                pi.rarity = item.rarity
                pc.inventory.append(pi)
            
            pg.members.append(pc)
        
        pw.guilds.append(pg)
    
    return pw


def encode_flatbuffers(builder, world):
    builder.Clear()
    
    guild_offsets = []
    for g in world.guilds:
        member_offsets = []
        for c in g.members:
            inv_offsets = []
            for item in c.inventory:
                name_off = builder.CreateString(item.name)
                rarity_off = builder.CreateString(item.rarity)
                FBItem.ItemStart(builder)
                FBItem.ItemAddId(builder, item.id)
                FBItem.ItemAddName(builder, name_off)
                FBItem.ItemAddValue(builder, item.value)
                FBItem.ItemAddWeight(builder, item.weight)
                FBItem.ItemAddRarity(builder, rarity_off)
                inv_offsets.append(FBItem.ItemEnd(builder))
            
            FBCharacter.CharacterStartInventoryVector(builder, len(inv_offsets))
            for off in reversed(inv_offsets):
                builder.PrependUOffsetTRelative(off)
            inv_vector = builder.EndVector()
            
            FBCharacter.CharacterStartSkillsVector(builder, len(c.skills))
            for s in reversed(c.skills):
                builder.PrependInt32(s)
            skills_vector = builder.EndVector()
            
            name_off = builder.CreateString(c.name)
            
            FBCharacter.CharacterStart(builder)
            FBCharacter.CharacterAddName(builder, name_off)
            FBCharacter.CharacterAddLevel(builder, c.level)
            FBCharacter.CharacterAddHp(builder, c.hp)
            FBCharacter.CharacterAddMp(builder, c.mp)
            FBCharacter.CharacterAddIsAlive(builder, c.is_alive)
            # Position struct — must be created inline during table construction
            FBCharacter.CharacterAddPosition(builder, FBVec3.CreateVec3(builder, c.position.x, c.position.y, c.position.z))
            FBCharacter.CharacterAddSkills(builder, skills_vector)
            FBCharacter.CharacterAddInventory(builder, inv_vector)
            member_offsets.append(FBCharacter.CharacterEnd(builder))
        
        FBGuild.GuildStartMembersVector(builder, len(member_offsets))
        for off in reversed(member_offsets):
            builder.PrependUOffsetTRelative(off)
        members_vector = builder.EndVector()
        
        name_off = builder.CreateString(g.name)
        desc_off = builder.CreateString(g.description)
        
        FBGuild.GuildStart(builder)
        FBGuild.GuildAddName(builder, name_off)
        FBGuild.GuildAddDescription(builder, desc_off)
        FBGuild.GuildAddMembers(builder, members_vector)
        guild_offsets.append(FBGuild.GuildEnd(builder))
    
    FBWorldState.WorldStateStartGuildsVector(builder, len(guild_offsets))
    for off in reversed(guild_offsets):
        builder.PrependUOffsetTRelative(off)
    guilds_vector = builder.EndVector()
    
    seed_off = builder.CreateString(world.seed)
    
    FBWorldState.WorldStateStart(builder)
    FBWorldState.WorldStateAddWorldId(builder, world.world_id)
    FBWorldState.WorldStateAddSeed(builder, seed_off)
    FBWorldState.WorldStateAddGuilds(builder, guilds_vector)
    root = FBWorldState.WorldStateEnd(builder)
    builder.Finish(root)
    return bytes(builder.Output())


def main():
    print("Generating Data...")
    world = create_benchmark_data()
    plain = to_dict(world)
    proto_world = create_proto_data(world)
    print("Data Generation Complete. Starting Benchmark...")
    
    ITERATIONS = 10

    # Prep sizes
    json_data = json.dumps(plain)
    msgpack_data = msgpack.packb(plain)
    bp_data = world.encode()
    proto_data = proto_world.SerializeToString()
    fb_builder = flatbuffers.Builder(65536)
    fb_data = encode_flatbuffers(fb_builder, world)
    
    print(f"\nPayload Sizes:")
    print(f"JSON: {len(json_data)}")
    print(f"MsgPack: {len(msgpack_data)}")
    print(f"BitPacker: {len(bp_data)}")
    print(f"Protobuf: {len(proto_data)}")
    print(f"FlatBuffers: {len(fb_data)}")

    # --- JSON ---
    start = time.time()
    for _ in range(ITERATIONS):
        s = json.dumps(plain)
        d = json.loads(s)
    elapsed = time.time() - start
    print(f"\nJSON:        {elapsed:.3f}s (Size: {len(json_data)} bytes)")

    # --- MsgPack ---
    start = time.time()
    for _ in range(ITERATIONS):
        b = msgpack.packb(plain)
        d = msgpack.unpackb(b)
    elapsed = time.time() - start
    print(f"MsgPack:     {elapsed:.3f}s (Size: {len(msgpack_data)} bytes)")

    # --- BitPacker ---
    start = time.time()
    for _ in range(ITERATIONS):
        b = world.encode()
        d = WorldState.decode(b)
    elapsed = time.time() - start
    print(f"BitPacker:   {elapsed:.3f}s (Size: {len(bp_data)} bytes)")

    # --- Protobuf ---
    start = time.time()
    for _ in range(ITERATIONS):
        b = proto_world.SerializeToString()
        d = ProtoWorldState()
        d.ParseFromString(b)
    elapsed = time.time() - start
    print(f"Protobuf:    {elapsed:.3f}s (Size: {len(proto_data)} bytes)")

    # --- FlatBuffers ---
    start = time.time()
    for _ in range(ITERATIONS):
        encode_flatbuffers(fb_builder, world)
    elapsed = time.time() - start
    print(f"FlatBuffers Encode: {elapsed:.3f}s (Size: {len(fb_data)} bytes)")

    start = time.time()
    for _ in range(ITERATIONS):
        w = FBWorldState.WorldState.GetRootAs(fb_data, 0)
        _ = w.WorldId()
        _ = w.GuildsLength()
    elapsed = time.time() - start
    print(f"FlatBuffers Decode: {elapsed:.3f}s")


if __name__ == "__main__":
    main()

File: benchmark/js/benchmark.js
const { WorldState, Guild, Character, Item, Vec3 } = require('../../generated/js/vec3');
const msgpack = require('msgpack-lite');
const protobuf = require('protobufjs');
const flatbuffers = require('flatbuffers');
const path = require('path');

function createBenchmarkData() {
    const world = new WorldState();
    world.world_id = 1;
    world.seed = "benchmark_seed";

    for (let g = 0; g < 1000; g++) {
        const guild = new Guild();
        guild.name = `Guild_${g}`;
        guild.description = "A very powerful guild";

        for (let c = 0; c < 20; c++) {
            const char = new Character();
            char.name = `Char_${g}_${c}`;
            char.level = (c + 1);
            char.hp = 100;
            char.mp = 50;
            char.is_alive = true;

            const pos = new Vec3();
            pos.x = 10;
            pos.y = 20;
            pos.z = 30;
            char.position = pos;

            char.skills = [1, 2, 3, 4, 5];

            for (let i = 0; i < 10; i++) {
                const item = new Item();
                item.id = (g * 1000 + c * 100 + i);
                item.name = "Item".repeat(50) + `_${g}_${c}_${i}`;
                item.value = i * 10;
                item.weight = 1;
                item.rarity = (i % 5 === 0) ? "Rare" : "Common";

                char.inventory.push(item);
            }

            guild.members.push(char);
        }

        world.guilds.push(guild);
    }

    world.loot_table = [];
    return world;
}

function toPlainObj(world) {
    return {
        world_id: world.world_id,
        seed: world.seed,
        guilds: world.guilds.map(g => ({
            name: g.name,
            description: g.description,
            members: g.members.map(c => ({
                name: c.name,
                level: c.level,
                hp: c.hp,
                mp: c.mp,
                is_alive: c.is_alive,
                position: { x: c.position.x, y: c.position.y, z: c.position.z },
                skills: c.skills,
                inventory: c.inventory.map(i => ({
                    id: i.id,
                    name: i.name,
                    value: i.value,
                    weight: i.weight,
                    rarity: i.rarity
                }))
            }))
        })),
        loot_table: []
    };
}

// FlatBuffers manual encode using the Builder API
function encodeFlatBuffers(builder, world) {
    builder.clear();

    const guildOffsets = [];
    for (const g of world.guilds) {
        const memberOffsets = [];
        for (const c of g.members) {
            // Inventory
            const invOffsets = [];
            for (const item of c.inventory) {
                const nameOff = builder.createString(item.name);
                const rarityOff = builder.createString(item.rarity);
                // Item table: id, name, value, weight, rarity
                builder.startObject(5);
                builder.addFieldInt32(0, item.id, 0);
                builder.addFieldOffset(1, nameOff, 0);
                builder.addFieldInt32(2, item.value, 0);
                builder.addFieldInt32(3, item.weight, 0);
                builder.addFieldOffset(4, rarityOff, 0);
                invOffsets.push(builder.endObject());
            }

            // Inventory vector
            builder.startVector(4, invOffsets.length, 4);
            for (let i = invOffsets.length - 1; i >= 0; i--) builder.addOffset(invOffsets[i]);
            const invVector = builder.endVector();

            // Skills vector
            builder.startVector(4, c.skills.length, 4);
            for (let i = c.skills.length - 1; i >= 0; i--) builder.addInt32(c.skills[i]);
            const skillsVector = builder.endVector();

            const nameOff = builder.createString(c.name);

            // Character table: name, level, hp, mp, is_alive, position, skills, inventory
            builder.startObject(8);
            builder.addFieldOffset(0, nameOff, 0);
            builder.addFieldInt32(1, c.level, 0);
            builder.addFieldInt32(2, c.hp, 0);
            builder.addFieldInt32(3, c.mp, 0);
            builder.addFieldInt8(4, c.is_alive ? 1 : 0, 0);
            // Position struct (Vec3: 3 x int32 = 12 bytes) — must be inline
            builder.prep(4, 12);
            builder.writeInt32(c.position.z);
            builder.writeInt32(c.position.y);
            builder.writeInt32(c.position.x);
            builder.addFieldStruct(5, builder.offset(), 0);
            builder.addFieldOffset(6, skillsVector, 0);
            builder.addFieldOffset(7, invVector, 0);
            memberOffsets.push(builder.endObject());
        }

        // Members vector
        builder.startVector(4, memberOffsets.length, 4);
        for (let i = memberOffsets.length - 1; i >= 0; i--) builder.addOffset(memberOffsets[i]);
        const membersVector = builder.endVector();

        const nameOff = builder.createString(g.name);
        const descOff = builder.createString(g.description);

        // Guild table: name, description, members
        builder.startObject(3);
        builder.addFieldOffset(0, nameOff, 0);
        builder.addFieldOffset(1, descOff, 0);
        builder.addFieldOffset(2, membersVector, 0);
        guildOffsets.push(builder.endObject());
    }

    // Guilds vector
    builder.startVector(4, guildOffsets.length, 4);
    for (let i = guildOffsets.length - 1; i >= 0; i--) builder.addOffset(guildOffsets[i]);
    const guildsVector = builder.endVector();

    const seedOff = builder.createString(world.seed);

    // WorldState table: world_id, seed, guilds, loot_table
    builder.startObject(4);
    builder.addFieldInt32(0, world.world_id, 0);
    builder.addFieldOffset(1, seedOff, 0);
    builder.addFieldOffset(2, guildsVector, 0);
    // empty loot_table - skip
    const root = builder.endObject();
    builder.finish(root);
    return builder.asUint8Array();
}

async function main() {
    console.log("Generating Data...");
    const data = createBenchmarkData();
    const plainData = toPlainObj(data);
    console.log("Data Generation Complete. Starting Benchmark...");

    const ITERATIONS = 10;

    // --- JSON ---
    {
        const start = process.hrtime.bigint();
        let size = 0;
        for (let i = 0; i < ITERATIONS; i++) {
            const encoded = JSON.stringify(plainData);
            if (i === 0) size = encoded.length;
            const decoded = JSON.parse(encoded);
        }
        const end = process.hrtime.bigint();
        const duration = Number(end - start) / 1e9;
        console.log(`JSON:        ${duration.toFixed(3)}s (Size: ${size} bytes)`);
    }

    // --- MsgPack ---
    {
        const start = process.hrtime.bigint();
        let size = 0;
        for (let i = 0; i < ITERATIONS; i++) {
            const encoded = msgpack.encode(plainData);
            if (i === 0) size = encoded.length;
            const decoded = msgpack.decode(encoded);
        }
        const end = process.hrtime.bigint();
        const duration = Number(end - start) / 1e9;
        console.log(`MsgPack:     ${duration.toFixed(3)}s (Size: ${size} bytes)`);
    }

    // --- BitPacker ---
    {
        try {
            const start = process.hrtime.bigint();
            let size = 0;
            for (let i = 0; i < ITERATIONS; i++) {
                const encoded = data.encode();
                if (i === 0) size = encoded.length;
                const decoded = WorldState.decode(encoded);
            }
            const end = process.hrtime.bigint();
            const duration = Number(end - start) / 1e9;
            console.log(`BitPacker:   ${duration.toFixed(3)}s (Size: ${size} bytes)`);
        } catch (e) {
            console.error("BitPacker Error:", e.message);
        }
    }

    // --- Protobuf ---
    {
        try {
            const root = await protobuf.load(path.join(__dirname, '../schemas/bench_complex.proto'));
            const ProtoWorldState = root.lookupType('bench_proto.WorldState');

            // Convert to proto-compatible object
            const protoObj = {
                worldId: plainData.world_id,
                seed: plainData.seed,
                guilds: plainData.guilds.map(g => ({
                    name: g.name,
                    description: g.description,
                    members: g.members.map(c => ({
                        name: c.name,
                        level: c.level,
                        hp: c.hp,
                        mp: c.mp,
                        isAlive: c.is_alive,
                        position: c.position,
                        skills: c.skills,
                        inventory: c.inventory.map(i => ({
                            id: i.id,
                            name: i.name,
                            value: i.value,
                            weight: i.weight,
                            rarity: i.rarity
                        }))
                    }))
                })),
                lootTable: []
            };

            const msg = ProtoWorldState.create(protoObj);

            const start = process.hrtime.bigint();
            let size = 0;
            for (let i = 0; i < ITERATIONS; i++) {
                const encoded = ProtoWorldState.encode(msg).finish();
                if (i === 0) size = encoded.length;
                const decoded = ProtoWorldState.decode(encoded);
            }
            const end = process.hrtime.bigint();
            const duration = Number(end - start) / 1e9;
            console.log(`Protobuf:    ${duration.toFixed(3)}s (Size: ${size} bytes)`);
        } catch (e) {
            console.error("Protobuf Error:", e.message);
        }
    }

    // --- FlatBuffers ---
    {
        try {
            const builder = new flatbuffers.Builder(65536);

            // Encode
            let fbBytes = encodeFlatBuffers(builder, data);
            const size = fbBytes.length;

            const start = process.hrtime.bigint();
            for (let i = 0; i < ITERATIONS; i++) {
                encodeFlatBuffers(builder, data);
            }
            const end = process.hrtime.bigint();
            const duration = Number(end - start) / 1e9;
            console.log(`FlatBuffers Encode: ${duration.toFixed(3)}s (Size: ${size} bytes)`);

            // Decode (access root fields)
            const buf = new flatbuffers.ByteBuffer(fbBytes);
            const start2 = process.hrtime.bigint();
            for (let i = 0; i < ITERATIONS; i++) {
                // FlatBuffers zero-copy: just reading offset
                const off = buf.readInt32(buf.position()) + buf.position();
                // Access world_id (field 0, vtable offset 4)
                const vtable = off - buf.readInt32(off);
                const vlen = buf.readInt16(vtable);
                if (vlen > 4) {
                    const _ = buf.readInt32(off + buf.readInt16(vtable + 4));
                }
            }
            const end2 = process.hrtime.bigint();
            const duration2 = Number(end2 - start2) / 1e9;
            console.log(`FlatBuffers Decode: ${duration2.toFixed(3)}s`);
        } catch (e) {
            console.error("FlatBuffers Error:", e.message);
        }
    }
}

main().catch(console.error);
